blockchain/v2/scheduler.go

```diff
@@ -0,0 +1,548 @@
+			min = height
+		}
+	}
+
+	return min
+}
+
+func (sc *schedule) pendingFrom(peerID p2p.ID) []int64 {
+	heights := []int64{}
+	for height, pendingPeerID := range sc.pendingBlocks {
+		if pendingPeerID == peerID {
+			heights = append(heights, height)
+		}
+	}
+	return heights
+}
+
+func (sc *schedule) selectPeer(peers []*scPeer) *scPeer {
+	// FIXME: properPeerSelector
+	s := rand.NewSource(time.Now().Unix())
+	r := rand.New(s)
+
+	return peers[r.Intn(len(peers))]
+}
+
+// XXX: this duplicates the logic of peersInactiveSince and peersSlowerThan
+func (sc *schedule) prunablePeers(peerTimout time.Duration, minRecvRate int64, now time.Time) []p2p.ID {
+	prunable := []p2p.ID{}
+	for peerID, peer := range sc.peers {
+		if now.Sub(peer.lastTouched) > peerTimout || peer.lastRate < minRecvRate {
+			prunable = append(prunable, peerID)
+		}
+	}
+
+	return prunable
+}
+
+func (sc *schedule) numBlockInState(targetState blockState) uint32 {
+	var num uint32 = 0
+	for _, state := range sc.blockStates {
+		if state == targetState {
+			num++
+		}
+	}
+	return num
+}
+
+type Scheduler struct {
+	sc             *schedule
+	targetPending  uint32 // the number of blocks we want in blockStatePending
+	targetReceived uint32 // the number of blocks we want in blockStateReceived
+	minRecvRate    int64
+	peerTimeout    time.Duration
+}
+
+func NewScheduler(minHeight int64, minRecvRate int64, peerTimeout time.Duration,
+	targetPending uint32, targetReceived uint32) *Scheduler {
+	return &Scheduler{
+		sc:             newSchedule(minHeight),
+		minRecvRate:    minRecvRate,
+		peerTimeout:    peerTimeout,
+		targetPending:  targetPending,
+		targetReceived: targetReceived,
+	}
+}
+
+type Skip struct {
+}
+
+func (sdr *Scheduler) handleAddPeer(peerID p2p.ID) Event {
+	err := sdr.sc.addPeer(peerID)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	return Skip{}
+}
+
+type Events []Event
+
+func (sdr *Scheduler) handleRemovePeer(peerID p2p.ID) Event {
+	err := sdr.sc.removePeer(peerID)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	return Skip{}
+}
+
+func (sdr *Scheduler) handleStatusResponse(peerID p2p.ID, height int64, now time.Time) Event {
+	err := sdr.sc.touchPeer(peerID, now)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	err = sdr.sc.setPeerHeight(peerID, height)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	return Skip{}
+}
+
+type bcBlockResponseEv struct {
+	peerID  p2p.ID
+	height  int64
+	block   *types.Block
+	msgSize int64
+}
+
+type scBlockReceivedEv struct {
+	peerID p2p.ID
+}
+
+func (sdr *Scheduler) handleBlockResponse(peerID p2p.ID, msg *bcBlockResponseEv, now time.Time) Event {
+	err := sdr.sc.touchPeer(peerID, now)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	err = sdr.sc.markReceived(peerID, msg.height, msg.msgSize, now)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	return scBlockReceivedEv{peerID}
```

Comment:
Who receives this event and wouldn't the height be needed?

Reply:
Good catch! Yes this will be sent to the processor and require the height, blockData, etc.